core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java

```diff
@@ -487,4 +339,163 @@ public boolean isFinished()
+            if (minMaxPossible) {
+                minMaxComparison = blockTypeOperators.getComparisonUnorderedLastOperator(type);
+            }
+            blockBuilder = type.createBlockBuilder(null, EXPECTED_BLOCK_BUILDER_SIZE);
+            valueSet = createUnboundedEqualityTypedSet(
+                    type,
+                    blockTypeOperators.getEqualOperator(type),
+                    blockTypeOperators.getHashCodeOperator(type),
+                    blockBuilder,
+                    EXPECTED_BLOCK_BUILDER_SIZE,
+                    format("DynamicFilterSourceOperator_%s_%d", planNodeId, channel.index));
+        }
+
+        public void process(Block block)
+        {
+            switch (state) {
+                case SET:
+                    for (int position = 0; position < block.getPositionCount(); ++position) {
+                        valueSet.add(block, position);
+                    }
+                    if (valueSet.getRetainedSizeInBytes() > maxFilterSizeInBytes || valueSet.size() > maxDistinctValues) {
+                        dfChannels--;
+                        if (minMaxPossible) {
+                            updateMinMaxValues(blockBuilder.build(), minMaxComparison);
+                            minMaxChannels++;
+                            state = ChannelState.MIN_MAX;
+                        }
+                        else {
+                            checkIfEmpty();
+                            state = ChannelState.NONE;
+                        }
+                        valueSet = null;
+                        blockBuilder = null;
+                    }
+                    break;
+                case MIN_MAX:
+                    updateMinMaxValues(block, minMaxComparison);
+                    break;
+                case NONE:
+                    break;
+            }
+        }
+
+        public Optional<Domain> getDomain()
+        {
+            Type type = channels.get(channelIndex).type;
+            return switch (state) {
+                case SET -> Optional.of(convertToDomain(type, blockBuilder.build()));
+                case MIN_MAX -> {
+                    if (minValues == null) {
+                        // all values were null
+                        yield Optional.of(Domain.none(type));
+                    }
+                    Object min = blockToNativeValue(type, minValues);
+                    Object max = blockToNativeValue(type, maxValues);
+                    yield Optional.of(Domain.create(ValueSet.ofRanges(range(type, min, true, max, true)), false));
+                }
+                case NONE -> Optional.empty();
+            };
+        }
+
+        public void disableMinMax()
+        {
+            minMaxPossible = false;
+            if (state == ChannelState.MIN_MAX) {
+                state = ChannelState.NONE;
+                minMaxChannels--;
+            }
+            checkIfEmpty();
```

Comment:
We don't need to call `checkIfEmpty` here, it can be done once after the for loop on `disableMinMax` in `addInput`